README.md
# knex-prometheus-exporter

[![Build Status](https://travis-ci.org/hekike/knex-prometheus-exporter.svg?branch=master)](https://travis-ci.org/hekike/knex-prometheus-exporter)

Prometheus exporter for [knex](https://www.npmjs.com/package/knex)

```sh
npm i -S knex-prometheus-exporter prom-client
```

## API

## knexExporter

Knex exporter

**Metrics:**

    # HELP knex_query_duration_seconds histogram of query responses
    # TYPE knex_query_duration_seconds histogram
    knex_query_duration_seconds_bucket{le="0.003"} 1
    knex_query_duration_seconds_bucket{le="0.03"} 2
    knex_query_duration_seconds_bucket{le="0.1"} 2
    knex_query_duration_seconds_bucket{le="0.3"} 2
    knex_query_duration_seconds_bucket{le="1.5"} 2
    knex_query_duration_seconds_bucket{le="10"} 2
    knex_query_duration_seconds_bucket{le="+Inf"} 2
    knex_query_duration_seconds_sum 0.021
    knex_query_duration_seconds_count 2

    # HELP knex_query_errors_total counter of query errors with labels: error
    # TYPE knex_query_errors_total counter
    knex_query_errors_total 0
    knex_query_errors_total{error="error message"} 1

**Parameters**

-   `knex` **Knex** knex instance
-   `opts` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)?** options
    -   `opts.register` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** register to use
    -   `opts.labels` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** extra labels (optional, default `{}`)
    -   `opts.prefix` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** metric name prefix (optional, default `"knex_"`)
    -   `opts.queryDurarionName` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** query duration metric name (histogram) (optional, default `"query_duration_seconds"`)
    -   `opts.responseTimeBuckets` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)>**  query duration buckets (optional, default `[0.003,0.03,0.1,0.3,1.5,10]`)
    -   `opts.queryErrorName` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** query errorr total name (counter) (optional, default `"query_errors_total"`)
    -   `opts.queryErrorWithErrorLabel` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** collect err.message as "error" label (optional, default `true`)

**Examples**

```javascript
const promClient = require('prom-client');
const Knex = require('knex');
const knexExporter = require('knex-prometheus-exporter');

const knex = Knex({
 client: 'mysql'
 ...
});
const exporter = knexExporter(knex);

console.log(promClient.register.metrics())
// =>
// # HELP knex_query_duration_seconds histogram of query responses
// # TYPE knex_query_duration_seconds histogram
// knex_query_duration_seconds_bucket{le="0.003"} 1
// knex_query_duration_seconds_bucket{le="0.03"} 2
// knex_query_duration_seconds_bucket{le="0.1"} 2
// knex_query_duration_seconds_bucket{le="0.3"} 2
// knex_query_duration_seconds_bucket{le="1.5"} 2
// knex_query_duration_seconds_bucket{le="10"} 2
// knex_query_duration_seconds_bucket{le="+Inf"} 2
// knex_query_duration_seconds_sum 0.021
// knex_query_duration_seconds_count 2

// # HELP knex_query_errors_total counter of query errors with labels: error
// # TYPE knex_query_errors_total counter
// knex_query_errors_total 0
// knex_query_errors_total{error="error message"} 1

// Unsubscribe:
exporter.off();
```

Returns **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** { off, registry }-   off: unscribe from metrics, registry: prom-client registry


CHANGELOG.md
<a name="3.0.0"></a>
## 3.0.0 (2022-12-19)

#### Breaking according the prom-client 13.x

changed: The following functions are now async (return a promise):
registry.metrics()
registry.getMetricsAsJSON()
registry.getMetricsAsArray()
registry.getSingleMetricAsString()


<a name="2.1.0"></a>
## 2.1.0 (2017-12-09)


#### Features

* **exporter:** support labels ([7de8ed68](git+https://github.com/hekike/knex-prometheus-exporter.git/commit/7de8ed68))


<a name="2.0.0"></a>
## 2.0.0 (2017-12-08)


#### Features

* **knex:** register support and prefix ([5a5af15b](git+https://github.com/hekike/knex-prometheus-exporter.git/commit/5a5af15b))


#### Breaking Changes

* external registry

 ([5a5af15b](git+https://github.com/hekike/knex-prometheus-exporter.git/commit/5a5af15b))


<a name="1.0.1"></a>
### 1.0.1 (2017-12-08)


#### Bug Fixes

* **knex:** prefix metrics with knex_ ([588b7f4d](git+https://github.com/hekike/knex-prometheus-exporter.git/commit/588b7f4d))


<a name="1.0.0"></a>
## 1.0.0 (2017-12-08)


#### Bug Fixes

* **exporter:** move prom-client to peer dependency ([31d97c21](git+https://github.com/hekike/knex-prometheus-exporter.git/commit/31d97c21))


#### Features

* **exporter:** implement ([150da51d](git+https://github.com/hekike/knex-prometheus-exporter.git/commit/150da51d))



test/e2e.test.js
const path = require('path');
const Knex = require('knex');
const promClient = require('prom-client');
const test = require('ava');
const dedent = require('dedent');
const knexExporter = require('../lib');

const knex = Knex({
  client: 'sqlite3',
  connection: {
    filename: path.join(__dirname, './mydb.sqlite')
  }
});
const tableName = 'users';
let exporter;
let register;

test.beforeEach(async () => {
  register = new promClient.Registry();

  await knex.schema.createTableIfNotExists(tableName, table => {
    table.increments();
    table.string('name').notNullable();
  });
});
test.afterEach.always(async () => {
  await knex(tableName).delete();
  exporter.off();
});

test.serial('measures query', async t => {
  exporter = await knexExporter(knex, {
    register,
    queryDurarionNameBuckets: [0.1, 0.3, 1.5, 10] // large enough buckets to avoid spreading
  });

  await knex(tableName).insert({
    name: 'Jane'
  });

  const user = await knex(tableName)
    .select('id', 'name')
    .first();

  const metricsOutput = await register.metrics();
  const [, sum] = metricsOutput.match(
    /knex_query_duration_seconds_sum ([0-9]+.[0-9]+)/
  );
  t.deepEqual(user.name, 'Jane');
  t.true(sum > 0, 'sum is larger than zero');
  t.deepEqual(
    metricsOutput,
    dedent`
      # HELP knex_query_duration_seconds histogram of query responses
      # TYPE knex_query_duration_seconds histogram
      knex_query_duration_seconds_bucket{le="0.1"} 2
      knex_query_duration_seconds_bucket{le="0.3"} 2
      knex_query_duration_seconds_bucket{le="1.5"} 2
      knex_query_duration_seconds_bucket{le="10"} 2
      knex_query_duration_seconds_bucket{le="+Inf"} 2
      knex_query_duration_seconds_sum ${sum}
      knex_query_duration_seconds_count 2

      # HELP knex_query_errors_total counter of query errors with labels: error
      # TYPE knex_query_errors_total counter\n
    `
  );
});

test.serial('counts query errors', async t => {
  t.plan(2);
  exporter = await knexExporter(knex, { register });

  let queryError;

  try {
    await knex(tableName).insert({
      invalid_field: 'Jane'
    });
  } catch (err) {
    queryError = err;
    t.truthy(err);
  }

  const metricsOutput = await register.metrics();

  t.deepEqual(
    metricsOutput,
    dedent`
      # HELP knex_query_duration_seconds histogram of query responses
      # TYPE knex_query_duration_seconds histogram
      knex_query_duration_seconds_bucket{le="0.003"} 0
      knex_query_duration_seconds_bucket{le="0.03"} 0
      knex_query_duration_seconds_bucket{le="0.1"} 0
      knex_query_duration_seconds_bucket{le="0.3"} 0
      knex_query_duration_seconds_bucket{le="1.5"} 0
      knex_query_duration_seconds_bucket{le="10"} 0
      knex_query_duration_seconds_bucket{le="+Inf"} 0
      knex_query_duration_seconds_sum 0
      knex_query_duration_seconds_count 0

      # HELP knex_query_errors_total counter of query errors with labels: error
      # TYPE knex_query_errors_total counter
      knex_query_errors_total{error="${queryError.message}"} 1\n
    `
  );
});

test.serial('counts query errors without label', async t => {
  t.plan(2);
  exporter = await knexExporter(knex, {
    register,
    queryErrorWithErrorLabel: false
  });

  try {
    await knex(tableName).insert({
      invalid_field: 'Jane'
    });
  } catch (err) {
    t.truthy(err);
  }

  const metricsOutput = await register.metrics();

  t.deepEqual(
    metricsOutput,
    dedent`
      # HELP knex_query_duration_seconds histogram of query responses
      # TYPE knex_query_duration_seconds histogram
      knex_query_duration_seconds_bucket{le="0.003"} 0
      knex_query_duration_seconds_bucket{le="0.03"} 0
      knex_query_duration_seconds_bucket{le="0.1"} 0
      knex_query_duration_seconds_bucket{le="0.3"} 0
      knex_query_duration_seconds_bucket{le="1.5"} 0
      knex_query_duration_seconds_bucket{le="10"} 0
      knex_query_duration_seconds_bucket{le="+Inf"} 0
      knex_query_duration_seconds_sum 0
      knex_query_duration_seconds_count 0

      # HELP knex_query_errors_total counter of query errors
      # TYPE knex_query_errors_total counter
      knex_query_errors_total 1\n
    `
  );
});

test.serial('supports prefix', async t => {
  t.plan(2);
  exporter = await knexExporter(knex, {
    register,
    prefix: 'foo_',
    queryErrorWithErrorLabel: false
  });

  try {
    await knex(tableName).insert({
      invalid_field: 'Jane'
    });
  } catch (err) {
    t.truthy(err);
  }

  const metricsOutput = await register.metrics();

  t.deepEqual(
    metricsOutput,
    dedent`
      # HELP foo_query_duration_seconds histogram of query responses
      # TYPE foo_query_duration_seconds histogram
      foo_query_duration_seconds_bucket{le="0.003"} 0
      foo_query_duration_seconds_bucket{le="0.03"} 0
      foo_query_duration_seconds_bucket{le="0.1"} 0
      foo_query_duration_seconds_bucket{le="0.3"} 0
      foo_query_duration_seconds_bucket{le="1.5"} 0
      foo_query_duration_seconds_bucket{le="10"} 0
      foo_query_duration_seconds_bucket{le="+Inf"} 0
      foo_query_duration_seconds_sum 0
      foo_query_duration_seconds_count 0

      # HELP foo_query_errors_total counter of query errors
      # TYPE foo_query_errors_total counter
      foo_query_errors_total 1\n
    `
  );
});

test.serial('supports labels', async t => {
  t.plan(2);
  exporter = await knexExporter(knex, {
    register,
    labels: { foo: 'bar' },
    queryErrorWithErrorLabel: false,
    queryDurarionNameBuckets: [0.1, 0.3, 1.5, 10] // large enough buckets to avoid spreading
  });

  try {
    await knex(tableName).insert({
      invalid_field: 'Jane'
    });
  } catch (err) {
    t.truthy(err);
  }

  await knex(tableName)
    .select('id', 'name')
    .first();

  const metricsOutput = await register.metrics();
  const [, sum] = metricsOutput.match(
    /knex_query_duration_seconds_sum{foo="bar"} ([0-9]+.[0-9]+)/
  );

  t.deepEqual(
    metricsOutput,
    dedent`
      # HELP knex_query_duration_seconds histogram of query responses
      # TYPE knex_query_duration_seconds histogram
      knex_query_duration_seconds_bucket{le="0.1",foo="bar"} 1
      knex_query_duration_seconds_bucket{le="0.3",foo="bar"} 1
      knex_query_duration_seconds_bucket{le="1.5",foo="bar"} 1
      knex_query_duration_seconds_bucket{le="10",foo="bar"} 1
      knex_query_duration_seconds_bucket{le="+Inf",foo="bar"} 1
      knex_query_duration_seconds_sum{foo="bar"} ${sum}
      knex_query_duration_seconds_count{foo="bar"} 1

      # HELP knex_query_errors_total counter of query errors
      # TYPE knex_query_errors_total counter
      knex_query_errors_total{foo="bar"} 1\n
    `
  );
});


lib/knex.js
const assert = require('assert');
const promClient = require('prom-client');

const NS_PER_SEC = 1e9;

/**
 * Convert hrtime to seconds
 *
 * @private
 * @function hrtimeToSeconds
 * @param {Array} hrtime - hrtime [0, 100]
 * @returns {Number} seconds
 */
function hrtimeToSeconds(hrtime) {
  const diffInNanoSecond = hrtime[0] * NS_PER_SEC + hrtime[1];
  const diffInSeconds = diffInNanoSecond / NS_PER_SEC;

  return diffInSeconds;
}

/**
 * Knex exporter
 *
 * **Metrics:**
 *
 * ```
 * # HELP knex_query_duration_seconds histogram of query responses
 * # TYPE knex_query_duration_seconds histogram
 * knex_query_duration_seconds_bucket{le="0.003"} 1
 * knex_query_duration_seconds_bucket{le="0.03"} 2
 * knex_query_duration_seconds_bucket{le="0.1"} 2
 * knex_query_duration_seconds_bucket{le="0.3"} 2
 * knex_query_duration_seconds_bucket{le="1.5"} 2
 * knex_query_duration_seconds_bucket{le="10"} 2
 * knex_query_duration_seconds_bucket{le="+Inf"} 2
 * knex_query_duration_seconds_sum 0.021
 * knex_query_duration_seconds_count 2

 * # HELP knex_query_errors_total counter of query errors with labels: error
 * # TYPE knex_query_errors_total counter
 * knex_query_errors_total 0
 * knex_query_errors_total{error="error message"} 1
 * ```
 *
 * @public
 * @function knexExporter
 * @param {Knex} knex - knex instance
 * @param {Object} [opts] - options
 * @param {String} [opts.register] - register to use
 * @param {Object} [opts.labels={}] - extra labels
 * @param {String} [opts.prefix="knex_"] - metric name prefix
 * @param {String} [opts.queryDurarionName="query_duration_seconds"]
 *  - query duration metric name (histogram)
 * @param {Number[]} [opts.responseTimeBuckets=[0.003, 0.03, 0.1, 0.3, 1.5, 10]]
 *  -  query duration buckets
 * @param {String} [opts.queryErrorName="query_errors_total"]
 *  - query errorr total name (counter)
 * @param {Boolean} [opts.queryErrorWithErrorLabel=true]
 *  - collect err.message as "error" label
 * @returns {Object} { off, registry }
 *  - off: unscribe from metrics, registry: prom-client registry
 * @example
 * const promClient = require('prom-client');
 * const Knex = require('knex');
 * const knexExporter = require('knex-prometheus-exporter');
 *
 * const knex = Knex({
 *  client: 'mysql'
 *  ...
 * });
 * const exporter = knexExporter(knex);
 *
 * console.log(promClient.register.metrics())
 * // =>
 * // # HELP knex_query_duration_seconds histogram of query responses
 * // # TYPE knex_query_duration_seconds histogram
 * // knex_query_duration_seconds_bucket{le="0.003"} 1
 * // knex_query_duration_seconds_bucket{le="0.03"} 2
 * // knex_query_duration_seconds_bucket{le="0.1"} 2
 * // knex_query_duration_seconds_bucket{le="0.3"} 2
 * // knex_query_duration_seconds_bucket{le="1.5"} 2
 * // knex_query_duration_seconds_bucket{le="10"} 2
 * // knex_query_duration_seconds_bucket{le="+Inf"} 2
 * // knex_query_duration_seconds_sum 0.021
 * // knex_query_duration_seconds_count 2

 * // # HELP knex_query_errors_total counter of query errors with labels: error
 * // # TYPE knex_query_errors_total counter
 * // knex_query_errors_total 0
 * // knex_query_errors_total{error="error message"} 1
 *
 * // Unsubscribe:
 * exporter.off();
 */
async function knexExporter(
  knex,
  {
    register = promClient.register,
    labels = {},
    prefix = 'knex_',
    // Query duration
    queryDurarionName = 'query_duration_seconds',
    queryDurarionNameBuckets = [0.003, 0.03, 0.1, 0.3, 1.5, 10],
    // Query error
    queryErrorName = 'query_errors_total',
    queryErrorWithErrorLabel = true
  } = {}
) {
  assert(knex, 'knex instance is required');

  const queryErrorLabels = [];
  let queryErrorHelp = 'counter of query errors';

  if (queryErrorWithErrorLabel) {
    queryErrorLabels.push('error');
    queryErrorHelp += ` with labels: ${queryErrorLabels.join(', ')}`;
  }

  const queries = new Map();
  const queryDurarionMetric = new promClient.Histogram({
    name: `${prefix}${queryDurarionName}`,
    help: 'histogram of query responses',
    labelNames: Object.keys(labels),
    buckets: queryDurarionNameBuckets,
    registers: []
  });
  const queryErrorMetric = new promClient.Counter({
    name: `${prefix}${queryErrorName}`,
    help: queryErrorHelp,
    buckets: queryDurarionNameBuckets,
    labelNames: queryErrorLabels.concat(Object.keys(labels)),
    registers: []
  });
  const labelValues = Object.values(labels);

  await register.registerMetric(queryDurarionMetric);
  await register.registerMetric(queryErrorMetric);

  function onQuery(query) {
    queries.set(query.__knexQueryUid, {
      startTime: process.hrtime()
    });
  }

  function onQueryResponse(response, query) {
    const { startTime } = queries.get(query.__knexQueryUid);
    queries.delete(query.__knexQueryUid);

    const diffInHrtime = process.hrtime(startTime);
    const diffInSeconds = hrtimeToSeconds(diffInHrtime);

    if (labelValues.length) {
      queryDurarionMetric.labels(labelValues).observe(diffInSeconds);
    } else {
      queryDurarionMetric.observe(diffInSeconds);
    }
  }

  function onQueryError(error, query) {
    queries.delete(query.__knexQueryUid);
    queryErrorMetric.inc(
      Object.assign(
        queryErrorWithErrorLabel ? { error: error.message } : {},
        labels
      )
    );
  }

  function off() {
    knex.removeListener('query', onQuery);
    knex.removeListener('query-response', onQueryResponse);
    knex.removeListener('query-error', onQueryError);
  }

  knex.on('query', onQuery);
  knex.on('query-response', onQueryResponse);
  knex.on('query-error', onQueryError);

  return {
    off
  };
}

module.exports = knexExporter;
